### DISCUSSION TITLE
How to get dynamic route params from Server Components?

### DISCUSSION THREAD

**Discussion by rinald-ta** [vercel/next.js#51818] [Help]
### Summary

With the new app router, dynamic parameters are provided to certain server components like `layout` and `page` through the `params` prop. For other server components used inside a layout or page, it appears that it is necessary to pass any dynamic params as props, since context doesn't work for server components.

I am also aware that there is a [useParams](https://nextjs.org/docs/app/api-reference/functions/use-params) hook, but that only works for client components.

Am I missing something here? Is there a cleaner way to provide dynamic params (or search params) to nested server components in a page? Is prop drilling the only option for now?

**Comment by icyJoseph** [comment]
Hi,

This is in the docs, and the solution is to just use JavaScript.

- https://nextjs.org/docs/getting-started/react-essentials#sharing-data-between-server-components

> Since Server Components are not interactive and therefore do not read from React state, you don't need React context to share data. Instead, you can use native JavaScript patterns for common data that multiple Server Components need to access. For example, a module can be used to share a database connection across multiple components:

Though it is not very easily found I guess.

So you could do for example:

```ts
// app/store.tsx
const createStore = () => {
  let data: string | null = null;

  const setData = (next: string) => {
    data = next;
  };

  const getData = () => {
    return data;
  };

  return { setData, getData };
};

export const store = createStore();
```

And then at say:

```tsx
// app/[foo]/[bar]/page.tsx

import { Outer } from "../../Outer";
import { store } from "../../store";

export default function Page({
  params,
}: {
  params: { foo: string; bar: string };
}) {
  const { foo, bar } = params;
  store.setData(`/${foo}/${bar}`);
  return <Outer />;
}
```

And then you'd have:

```tsx
import { Inner } from "./Inner";

export const Outer = () => {
  return <Inner />;
};
```

Where, finally, Inner, just reads the data:

```tsx
import { store } from "./store";

export const Inner = () => {
  return <div>{store.getData()}</div>;
};
```

**Reply by VaghikSoghomonyan** [reply]
Hey, 
Thanks for sharing this! It’s a cool way to handle data between server components. But I’m a bit worried about what happens when two users hit the server at the same time. Looks like there could be a race condition with the shared store module where one user’s data might overwrite the other's. Since the store isn’t tied to individual requests, data could get mixed up, right?

Is there a way to tie the data store to each request to avoid this? Or maybe there's another workaround to keep the data straight when multiple users are making requests at the same time?

**Reply by icyJoseph** [reply]
Hi,

Yeah, someone else already run into that issue. And they shared a package they used to get around it, let me find it for you.

**Reply by icyJoseph** [reply]
Here you go: 

- https://github.com/manvalls/server-only-context#readme

For context on the issue: https://github.com/vercel/next.js/discussions/50405

**Reply by ludo237** [reply]
Try to avoid unnecessary store as much as you can as a good practice

**Reply by eric-burel** [reply]
The store described in this post is not user specific and currently corresponds to Next.js `unstable_cache` basically. To avoid props drilling it's more usual to rely on React `cache` which is scoped to the current user request (= one server render), which powers the server-only-context lib shared just a few comments above.

**Comment by smail-boundaoui** [comment]
`export default function ServerComponentPage({ params, searchParams }) {`
`  console.log(params)`
`  console.log(searchParams)`
`  return`
`}`

**Reply by smail-boundaoui** [reply]
[https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes#optional-catch-all-segments](https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes#optional-catch-all-segments)

**Reply by Basovs** [reply]
I didnt understood this solution provided by @smail-boundaoui ... Can some1 explain?

**Reply by Muhammad-Umar-Khan** [reply]
@Basovs what you need help with?

**Reply by zjy365** [reply]
Yes, this is correct, the reason is here https://nextjs.org/docs/app/api-reference/file-conventions/page

**Reply by anotheroneofthese** [reply]
You might miss it in the documentation smail-boundaoui linked, but in later versions of NextJS you need to specify that it is a `Promise`.
For example, using an API route to the page `/api/[id]/details`, you might want:

```typescript
export async function GET(
  _: NextRequest,
  { params }: { params: Promise<{ id: string }> },
  { searchParams }: { searchParams: Promise<{ id: string }> }
) {
  // rest of your code
}
```

**Comment by fallenleavesguy** [comment]
by the way. if you are using next-intl, you can get locale by using import { useLocale } from "next-intl";

**Reply by Eunit99** [reply]
How to use?

```JavaScript
import { useLocale } from "next-intl";

...

  console.log(useLocale()); // en

```

**Comment by rechenberger** [comment]
Hey everybody,

just built a library to handle this problem:
https://www.npmjs.com/package/@sodefa/next-server-context
It uses [`AsyncLocalStorage`](https://nodejs.org/api/async_context.html)

---

How to use:

/app/[myParam]/page.tsx

```tsx
import { pageContext } from '@sodefa/next-server-context'
import { NestedServerComponent } from './NestedServerComponent'

export default pageContext.Wrapper(({ params, searchParams }) => {
  return (
    <>
      <h1>MyPage</h1>
      <NestedServerComponent />
    </>
  )
})
```

/app/[myParam]/NestedServerComponent.tsx

```tsx
import { pageContext } from '@sodefa/next-server-context'

export const NestedServerComponent = () => {
  const { params, searchParams } = pageContext.getOrThrow()
  return <div>NestedServerComponent: {params.myParam}</div>
}
```

**Reply by eric-burel** [reply]
Did you benchmark this? What bothers me is that you are creating an AsyncLocalStorage, on top of the one already created by Next.Js, so you may pay an overhead price twice. Using React cache lets you use the AsyncLocalStorage that Next.js already creates for you instead of writing your own, I would favour that ([server-only-context approach](https://github.com/manvalls/server-only-context#readme)). But a benchmark would help telling if I am wrong or not.
Your tool might be cool to setup observability though.

**Reply by DenisBessa** [reply]
That's a very cool idea and library. But I'm worried about the implications of using it for PPR and other Next.js optimizations. Does anybody has some info about it?

**Comment by mercurydiogo** [comment]
Is there a native Next.js solution for this yet? If not, what might be the reason it hasn't been done? Feels like a pretty common problem in my mind

**Reply by shanehoban** [reply]
Yes, see a couple of comments above: https://github.com/vercel/next.js/discussions/51818#discussioncomment-7683980

**Reply by mercurydiogo** [reply]
This is for a page though, not a component. Right?

**Reply by shanehoban** [reply]
Yes but pages/routes are where these parameters come from, so you can pass them to child components of the page that are still server rendered, or use hooks (when unable or not within scope for whatever reason) in a client component.

**Reply by shanehoban** [reply]
To clarify, a standalone server component would be unaware of route parameters without being passed them specifically from the page/route that they originate from.

**Reply by mercurydiogo** [reply]
> Yes but pages/routes are where these parameters come from, so you can pass them to child components of the page that are still server rendered, or use hooks (when unable or not within scope for whatever reason) in a client component.

Yes, this is what is in the docs. But why shouldnt you be able to use conditional logic on the parameter (in a server component) without prop drilling just like usual useParams.

I don't understand why this is a technical challenge for nextjs

I'm sure I'm just missing something fundamental, but I don't understand what.

If anyone could elucidate me, I'd be much appreciated

**Comment by laurent512** [comment]
Any update on that ? Quite important to me to get dynamic url segment inside server components without deep props drilling :/

**Reply by 0-don** [reply]
yeah still waiting

**Reply by devajakamerus** [reply]
You could simply reuse the asynclocalstorage that the framework uses internally and access request details thru it. Similar to how cookies() and headers() work. Probably not the best idea but works fine, this definitely should be builtin.

`import { staticGenerationAsyncStorage } from "next/dist/client/components/static-generation-async-storage.external";

export default function Test() {
    const store = staticGenerationAsyncStorage.getStore();
    if (!store) {
        return null; // ...
    }
    return (
        < main>
            <p>Request url was : {store.urlPathname}</p>
        < /main>
    );
}
`
E: i don't know why the formattin just breaks, well you get the idea

**Comment by indicozy-eyewa** [comment]
A better approach I see is to use headers in middleware:
```js
// middleware.js
  const res = NextResponse.redirect(url);
    res.headers.set('x-store', storeCode);
    return res
```

```js
// server side
  const headerModule = require('next/headers');
  return headerModule.headers().get('x-store');
```

**Comment by mhmdAljefri** [comment]
I prefer using cookies in middleware.ts and then get value of it there for themes and local as well. but it can be used as you need.

```ts
// middleware.ts
export function middleware(request: NextRequest) {
  const response = NextResponse.next();
  response.cookies.set('lang',getLocalByRoute );

  return response
};
```

and in SSC
```ts
import { cookies } from "next/headers";

export default function ButtonWithTranslations() {
  const lang = cookies().get("lang")?.value || "en";

  // rest code
}
```

**Comment by EricWVGG** [comment]
I'm trying so hard to be an optimist about this whole "server components" thing, and then I crash into a problem so dumb and obvious… 

I absolutely cannot believe the App Router got pushed out the door without this sort of basic functionality. Every solution I've seen on this page makes sense but is ultimately a creaky hack over something that is just fundamental to the very concept of a server component. Augh.

**Reply by DrPye** [reply]
Using default javascript patterns instead is not a creaky hack. 

Just move back to the pages router until your more comfortable with RSC
